src/context/mod.rs

```diff
@@ -3,58 +3,88 @@ use std::{
     cell::{Ref, RefCell},
     ops,
     path::Path,
+    sync::Arc,
 };
 
-use derive_builder::Builder;
-use squalid::{IsEmpty, OptionExt};
-use tree_sitter_grep::{streaming_iterator::StreamingIterator, RopeOrSlice, SupportedLanguage};
+use tree_sitter_grep::{
+    streaming_iterator::StreamingIterator, tree_sitter::Tree, RopeOrSlice, SupportedLanguage,
+};
 
 mod backward_tokens;
+mod fix;
 mod get_tokens;
+mod skip_options;
 
-use get_tokens::get_tokens;
+use backward_tokens::{get_backward_tokens, get_tokens_before_node};
+pub use fix::{Fixer, PendingFix};
+use get_tokens::{get_tokens, get_tokens_after_node};
+pub use skip_options::{SkipOptions, SkipOptionsBuilder};
 
-use self::{backward_tokens::get_backward_tokens, get_tokens::get_tokens_after_node};
 use crate::{
     rule::InstantiatedRule,
     tree_sitter::{Language, Node, Query},
     violation::{Violation, ViolationWithContext},
-    Config,
+    AggregatedQueries, Config,
 };
 
-pub struct QueryMatchContext<'a> {
+#[derive(Copy, Clone)]
+pub struct FileRunContext<'a> {
     pub path: &'a Path,
     pub file_contents: RopeOrSlice<'a>,
-    pub rule: &'a InstantiatedRule,
-    config: &'a Config,
+    pub tree: &'a Tree,
+    pub config: &'a Config,
     pub language: SupportedLanguage,
-    pending_fixes: RefCell<Option<Vec<PendingFix>>>,
-    pub violations: RefCell<Option<Vec<ViolationWithContext>>>,
+    pub(crate) aggregated_queries: &'a AggregatedQueries<'a>,
+    pub(crate) query: &'a Arc<Query>,
+    pub(crate) instantiated_rules: &'a [InstantiatedRule],
```

Comment:
Some of these things make sense to be visible to rule "consumers" and some of them are more just "implementation details" that should be hidden but are useful to "bundle up" here